### PR TITLE
Adaptive step improvement

### LINKED ISSUE #51: Improve adaptive curvature step in moment-curvature analysis
@Agent6-6-6:
> I've noticed as well that the way the moment curvature analysis adaptively scales the curvature step it almost always just keeps increasing the curvature in each step. so by the time you get closer to the failure point the answers get very widely spaced.

```
if moment_diff <= delta_m_min:
                        kappa_inc *= 2
                    elif moment_diff >= delta_m_max:
                        kappa_inc *= 0.5
```

>To demonstrate, first image is replacing 2 with 1.01, and 0.5 with 0.99 to arbitrarily limit to a much smaller step. You can appreciate the differences in the results in the below images for the same section. So by default the answer gets cruder and crude as it proceeds and you start to miss the fundamental behaviour I think. Note the general shape and higher curvature achieved when the steps are smaller.
Figure_1
Figure_2

> It feels like the final point should always be undertaken at the failure strain. At the moment the final analysis point returned is only the point evaluated before a moment step goes over the failure strain. Which with a cruder step could be quite a bit lower than the actual true failure strain.

_Originally posted by @Agent6-6-6 in https://github.com/robbievanleeuwen/concrete-properties/issues/44#issuecomment-1256912356_

### PR DESCRIPTION
**Pull request by robbievanleeuwen**
Fixes #51

### CODE REVIEW

**Comment by robbievanleeuwen** [comment]
@Agent6-6-6 if you have a chance, wondering if you'd mind having a quick look at this tosee what you think before I merge.

Cheers!

**Comment by robbievanleeuwen** [comment]
FYI there's a test documentation build [here](https://robbievanleeuwen.github.io/concrete-properties/test/).

P.S. the addition of the failure point was a very good suggestion, I think it works well!

**Review by Agent6-6-6 — CHANGES_REQUESTED** [review]
@robbievanleeuwen take a look at the comments and see what you think!

Note I have not run any examples but the code appears to address all of the points that were raised in #51 and gives good granular control over the resulting curve if the defaults are not cutting it!

**Comment by Agent6-6-6** [comment]
@robbievanleeuwen, wonder as well in the notebook example, perhaps an example illustrating the use of these new parameters to tune the accuracy of the generated curve? 

To show for example a default analysis vs one that is a bit more refined, perhaps demonstrating the tradeoff in terms of accuracy vs analysis time. But ultimately shows that limiting curvature is the same if that's all the user is interested in.

File: concreteproperties/concrete_section.py [hunk 1]
```diff
@@ -700,6 +761,15 @@ def service_normal_force_convergence(
                 moment_curvature._failure = True
                 moment_curvature.failure_geometry = meshed_geom
```
[the remark this answers is not in the document]
Comment (robbievanleeuwen) [reply concreteproperties/concrete_section.py:762]: ``service_normal_force_convergence()`` is also run by the ``brentq`` algorithm when finding the failure curvature, so this should update the ``failure_geometry`` correctly I believe.

The last ``service_normal_force_convergence()`` call should result in ``_failure_convergence`` being *almost* equal to one and during this call the ``failure_geometry`` is assigned so this should work.

File: concreteproperties/concrete_section.py [hunk 2]
```diff
@@ -740,6 +816,7 @@ def service_normal_force_convergence(
         moment_curvature._n_i = n
         moment_curvature._m_x_i = m_x
         moment_curvature._m_y_i = m_y
+        moment_curvature._failure_convergence = failure_convergence
 
         # calculate convergence
```
[the remark this answers is not in the document]
Comment (robbievanleeuwen) [reply concreteproperties/concrete_section.py:821]: Agreed.

File: concreteproperties/concrete_section.py [hunk 3]
```diff
@@ -590,8 +604,54 @@ def mcurve(kappa_inc=kappa_inc, progress=None):
...
+            moment_curvature.kappa.append(kappa)
+            moment_curvature.n.append(moment_curvature._n_i)
+            moment_curvature.m_x.append(moment_curvature._m_x_i)
+            moment_curvature.m_y.append(moment_curvature._m_y_i)
+            moment_curvature.m_xy.append(m_xy)
+            moment_curvature.convergence.append(moment_curvature._failure_convergence)
+
         # create progress bar
         if progress_bar:
             # create progress bar
```
[the remark this answers is not in the document]
Comment (robbievanleeuwen) [reply concreteproperties/concrete_section.py:657]: Fixed!

File: concreteproperties/results.py [hunk 4]
```diff
@@ -418,12 +418,14 @@ class MomentCurvatureResults:
     m_y: List[float] = field(default_factory=list)
     m_xy: List[float] = field(default_factory=list)
     failure_geometry: CPGeom = field(init=False)
+    convergence: List[float] = field(default_factory=list)
```
[the remark this answers is not in the document]
Comment (robbievanleeuwen) [reply concreteproperties/results.py:421]: Added.

File: concreteproperties/concrete_section.py [hunk 5]
```diff
@@ -590,8 +604,54 @@ def mcurve(kappa_inc=kappa_inc, progress=None):
...
                     iter += 1
 
+            # find kappa corresponding to failure strain:
+            # curvature before and after failure
+            kappa_a = moment_curvature.kappa[-1]
+            kappa_b = kappa
+
+            # this method (given a kappa) outputs the failure convergence
+            # (normalised to zero)
+            def failure_kappa(kappa_fail):
```
[the remark this answers is not in the document]
Comment (robbievanleeuwen) [reply concreteproperties/concrete_section.py:619]: See other comment, I think this is captured as ``service_normal_force_convergence`` is run in this.

**Comment by robbievanleeuwen** [comment]
Example modified, see [here](https://robbievanleeuwen.github.io/concrete-properties/test/notebooks/moment_curvature.html#Finetuning-Analysis-Parameters). Didn't want to make it too complex as the docs are already taking a far bit of time to build 😆

File: concreteproperties/concrete_section.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Agent6-6-6) [reply concreteproperties/concrete_section.py:762]: All good then if that's the case.

**Pull request merged**